csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -56,16 +56,16 @@ public abstract class BuilderBase<TArray, TBuilder> : IArrowArrayBuilder<byte, T
             protected TBuilder Instance => this as TBuilder;
             protected ArrowBuffer.Builder<int> ValueOffsets { get; }
             protected ArrowBuffer.Builder<byte> ValueBuffer { get; }
-            protected BooleanArray.Builder ValidityBuffer { get; }
+            protected ArrowBuffer.BitPackedBuilder ValidityBuffer { get; }
             protected int Offset { get; set; }
-            protected int NullCount { get; private set; }
+            protected int NullCount => this.ValidityBuffer.CountUnsetBits();
```

Comment:
Since `NullCount` is a property, it should have a fast runtime access performance (similar to accessing a field). However, this change is calling a method which is going to grow in time as the buffer grows. I think we should keep this a property with a backing field, so callers don't get into a performance trap where they are calling `NullCount` too often thinking it has fast access (since it is a property).

Reply:
This was slightly more involved to change, but now done: the `BitmapBuilder` class keeps track of the set/unset bit count during its operations.  This required adding a new method to `BitUtility`, along with appropriate unit tests.

I also took the opportunity to add verification of the new `BitmapBuilder.Num[Set|Unset]Bits` properties in all the unit tests.